### Tidy Data Exercises

**1.  Attendance Data**

**Load the ```attendance.csv``` file and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.**

**You should end up with something like this:**

**```
name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64```**

In [40]:
import pandas as pd

attendance = pd.read_csv('attendance.csv')
attendance

attendance.rename(columns={'Unnamed: 0':'name'}, inplace=True)
attendance.head(2)

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T

In [41]:
# restructure data using 'melt'
attendance_melt = attendance.melt(id_vars=['name'], var_name='date', value_name='score')
attendance_melt.head(2)


name        date score
0  Sally  2018-01-01     P
1   Jane  2018-01-01     A

In [42]:
attendance_melt = attendance_melt.replace('P',1).replace('A',0).replace('T',.9).replace('H',.5)
attendance_melt.head(2)

name        date  score
0  Sally  2018-01-01    1.0
1   Jane  2018-01-01    0.0

In [43]:
attendance_tidy = attendance_melt.pivot_table(index = 'name')
attendance_tidy

score
name         
Billy  0.5250
Jane   0.6875
John   0.9125
Sally  0.7625

**2. Coffee Levels**

**a. Read the ```coffee_levels.csv``` file.**


In [31]:
coffee = pd.read_csv('coffee_levels.csv')
coffee.head()


coffee.rename(columns={'coffee_carafe':'carafe'}, inplace=True)
coffee.head(2)

hour carafe  coffee_amount
0     8      x       0.816164
1     9      x       0.451018

**b. Transform the data so that each carafe is in it's own column.**


In [32]:
tidy_coffee = coffee.pivot(index = 'hour', columns = 'carafe', values = 'coffee_amount')
tidy_coffee

carafe         x         y         z
hour                                
8       0.816164  0.189297  0.999264
9       0.451018  0.521502  0.915990
10      0.843279  0.023163  0.144928
11      0.335533  0.235529  0.311495
12      0.898291  0.017009  0.771947
13      0.310711  0.997464  0.398520
14      0.507288  0.058361  0.864464
15      0.215043  0.144644  0.436364
16      0.183891  0.544676  0.280621
17      0.391560  0.594126  0.436677

**c. Is this the best shape for the data?**

In [ ]:
yes

**3.  Cake Recipes**

**a. Read the ```cake_recipes.csv``` data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.**


In [7]:
recipes = pd.read_csv('cake_recipes.csv')
recipes.rename(columns={'recipe:position':'recipe_position'}, inplace=True)
recipes


recipe_position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

**b. Tidy the data as necessary.**


In [8]:
recipes = pd.melt(recipes, id_vars = ['recipe_position'], var_name = 'temp', value_name = 'rating')

recipes.head(2)

recipe_position temp     rating
0        a:bottom  225  61.738655
1           a:top  225  51.709751

In [11]:
rec_pos = recipes.recipe_position.str.split(':', expand = True)
rec_pos.head(5)

0       1
0  a  bottom
1  a     top
2  b  bottom
3  b     top
4  c  bottom

In [14]:
recipes[['recipe','position']] = recipes.recipe_position.str.split(':', expand = True)
recipes.head(2)


recipe_position temp     rating recipe position
0        a:bottom  225  61.738655      a   bottom
1           a:top  225  51.709751      a      top

In [18]:
# recipes.drop(columns = 'recipe_position', inplace = True)
recipes.head(2)

temp     rating recipe position
0  225  61.738655      a   bottom
1  225  51.709751      a      top

**c. Which recipe, on average, is the best?** recipe b


In [19]:
recipes.groupby('recipe').rating.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: rating, dtype: float64

**d. Which oven temperature, on average, produces the best results?** 275


In [21]:
recipes.groupby('temp').rating.mean()

temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: rating, dtype: float64

**e. Which combination of recipe, rack position, and temperature gives the best result?** recipe b, bottom rack, 300 degrees

In [45]:
#recipes.sort_values(by = 'rating').tail(3)
recipes.nlargest(1, 'rating')

temp     rating recipe position
26  300  99.248541      b   bottom